GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs

```diff
@@ -32,6 +35,59 @@ public void ChangeMode(string path, int mode)
         public bool TryGetNormalizedPath(string path, out string normalizedPath, out string errorMessage)
         {
             return WindowsFileSystem.TryGetNormalizedPathImplementation(path, out normalizedPath, out errorMessage);
+        }
+
+        public bool HydrateFile(string fileName, byte[] buffer)
+        {
+            return NativeFileReader.TryReadFirstByteOfFile(fileName, buffer);
+        }
+
+        private class NativeFileReader
+        {
+            private const uint GenericRead = 0x80000000;
+            private const uint OpenExisting = 3;
+
+            public static bool TryReadFirstByteOfFile(string fileName, byte[] buffer)
+            {
+                using (SafeFileHandle handle = Open(fileName))
+                {
+                    if (!handle.IsInvalid)
+                    {
+                        return ReadOneByte(handle, buffer);
+                    }
+                }
+
+                return false;
+            }
+
+            private static SafeFileHandle Open(string fileName)
+            {
+                return CreateFile(fileName, GenericRead, (uint)(FileShare.ReadWrite | FileShare.Delete), 0, OpenExisting, 0, 0);
+            }
+
+            private static unsafe bool ReadOneByte(SafeFileHandle handle, byte[] buffer)
+            {
+                int bytesRead = 0;
+                return ReadFile(handle, buffer, 1, &bytesRead, 0);

```

Comment:
Why are you passing in `&bytesRead`?  If you just pass in `bytesRead` there should be no need to mark this `unsafe`.

Reply:
Oh sorry, I missed that this was for `int* numberOfBytesRead`

Reply:
If you change the `DllImport` from:

`[Out] int* numberOfBytesRead,`

to 

`ref int numberOfBytesRead,`

You can drop the `unsafe`

Reply:
See example here (the second parameter)

https://docs.microsoft.com/en-us/dotnet/framework/interop/marshaling-different-types-of-arrays

*C API*
```
int TestRefArrayOfInts(int** ppArray, int* pSize);  
```

*DllImport in C#*
```
   [ DllImport( "..\\LIB\\PinvokeLib.dll" )]
   public static extern int TestRefArrayOfInts( ref IntPtr array, 
      ref int size );
```

Reply:
Thanks!